dispatcher_startup.go

```diff
@@ -0,0 +1,269 @@
+}
+
+func (s *PhasedStarter) abort(errs []error) error {
+	// Failed to start so stop everything that was started.
+	wait := errorsync.ErrorWaiter{}
+	s.startedMu.Lock()
+	for _, lc := range s.started {
+		wait.Submit(lc.Stop)
+	}
+	s.startedMu.Unlock()
+	if newErrors := wait.Wait(); len(newErrors) > 0 {
+		errs = append(errs, newErrors...)
+	}
+
+	return multierr.Combine(errs...)
+}
+
+func (s *PhasedStarter) setRouters() {
+	// Don't need synchronization, since we always call this in a lifecycle.Once
+	// in the dispatcher.
+	s.log.Debug("setting router for inbounds")
+	for _, ib := range s.dispatcher.inbounds {
+		ib.SetRouter(s.dispatcher.table)
+	}
+	s.log.Debug("set router for inbounds")
+}
+
+// PhasedStopper is a more granular alternative to the Dispatcher's all-in-one
+// Stop method. Rather than stopping the inbounds, outbounds, and transports
+// in one call, it lets the user choose when to trigger each phase of
+// dispatcher shutdown. For details on the interaction of Stop and phased
+// shutdown, see the documentation for the Dispatcher's PhasedStop method.
+//
+// The user of a PhasedStopper is responsible for correctly ordering shutdown:
+// inbounds MUST be stopped before outbounds, which MUST be stopped before
+// transports. Attempting shutdown in any other order will return an error.
+type PhasedStopper struct {
+	dispatcher *Dispatcher
+	log        *zap.Logger
+
+	inboundsStopInitiated   atomic.Bool
+	inboundsStopped         atomic.Bool
+	outboundsStopInitiated  atomic.Bool
+	outboundsStopped        atomic.Bool
+	transportsStopInitiated atomic.Bool
+}
+
+// StopInbounds is the first step in shutdown. It stops all inbounds
+// configured on the dispatcher, which stops routing RPCs to all registered
+// procedures. It's safe to call concurrently, but all calls after the first
+// return an error.
+func (s *PhasedStopper) StopInbounds() error {
+	if s.inboundsStopInitiated.Swap(true) {
+		return errors.New("already began stopping inbounds")
+	}
+	defer s.inboundsStopped.Store(true)
+	s.log.Debug("stopping inbounds")
+	wait := errorsync.ErrorWaiter{}
+	for _, ib := range s.dispatcher.inbounds {
+		wait.Submit(ib.Stop)
+	}
+	if errs := wait.Wait(); len(errs) > 0 {
+		return multierr.Combine(errs...)
+	}
+	s.log.Debug("stopped inbounds")
+	return nil
+}
+
+// StopOutbounds is the second step in shutdown. It stops all outbounds
+// configured on the dispatcher, which stops clients from making outbound
+// RPCs. It's safe to call concurrently, but all calls after the first return
+// an error.
+func (s *PhasedStopper) StopOutbounds() error {
+	if !s.inboundsStopped.Load() {
+		return errors.New("must stop inbounds first")
+	}
+	if s.outboundsStopInitiated.Swap(true) {
+		return errors.New("already began stopping outbounds")
+	}
+	defer s.outboundsStopped.Store(true)
+	s.log.Debug("stopping outbounds")
+	wait := errorsync.ErrorWaiter{}
+	for _, o := range s.dispatcher.outbounds {
+		if o.Unary != nil {
+			wait.Submit(o.Unary.Stop)
```

Comment:
Is there ever a possibility of a non-nil outbound with a nil `Stop` method? If so, we might want to add an additional nil check in the conditional since a call to `wait.Submit` will panic with a `nil` arg.

Reply:
Not as far as I know. `o.Unary.Stop` isn't a field, it's a method pointer - it can't be nil. Since we're checking for nil receivers above, we should be safe.